aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -591,16 +639,25 @@ Tensor scaled_dot_product_attention(
       bool compute_logsumexp =
           (query_.requires_grad() || key.requires_grad() ||
            value.requires_grad());
+      if (attn_mask.has_value()) {
+        // Expand to 4d case
+        attn_mask = attn_mask.value().expand_symint(
```

Comment:
Random thought, but I saw that the kernel takes into account the strides of the bias. Is it really always better to just expand for the 2D case or is it maybe better to also add a `.contiguous` after? I could see a small Tensor being this much in demand being harmful to performance.

At the same time, it saves a ton of memory to not replicate the mask for each batch entry. I'm mostly curious, but I could see that maybe in some cases it'd be better to replicate the mask per head at least? Might be worth some follow up investigation / performance tuning.